shim/shim.c

```diff
@@ -0,0 +1,621 @@
+	if ( errno || *endptr ) {
+		return -1;
+	}
+	return num;
+}
+
+/*!
+ * Print program usage
+ */
+void
+print_usage(void) {
+	printf("%s: Usage\n", program_name);
+        printf("  -c,  --container-id   Container id\n");
+        printf("  -p,  --proxy-sock-fd  File descriptor of the socket connected to cc-proxy\n");
+        printf("  -o,  --proxy-io-fd    File descriptor of I/0 fd sent by the cc-proxy\n");
+        printf("  -s,  --seq-no         Sequence no for stdin and stdout\n");
+        printf("  -e,  --err-seq-no     Sequence no for stderr\n");
+        printf("  -d,  --debug          Enable debug output\n");
+        printf("  -h,  --help           Display this help message\n");
+}
+
+int
+main(int argc, char **argv)
+{
+	struct cc_shim shim = {
+		.container_id   =  NULL,
+		.proxy_sock_fd  = -1,
+		.proxy_io_fd    = -1,
+		.io_seq_no      =  0,
+		.err_seq_no     =  0,
+	};
+	struct             sockaddr_un proxy_address;
+	char              *container_id = NULL;
+	char              *proxy_sock_path;
+	struct pollfd      poll_fds[MAX_POLL_FDS] = {{-1}};
+	nfds_t             nfds = 0;
+	int                ret;
+	struct sigaction   sa;
+	int                c;
+	bool               debug = false;
+
+	program_name = argv[0];
+
+	struct option prog_opts[] = {
+		{"container-id", required_argument, 0, 'c'},
+		{"proxy-sock-fd", required_argument, 0, 'p'},
+		{"proxy-io-fd", required_argument, 0, 'o'},
+		{"seq-no", required_argument, 0, 's'},
+		{"err-seq-no", required_argument, 0, 'e'},
+		{"debug", no_argument, 0, 'd'},
+		{"help", no_argument, 0, 'h'},
+		{ 0, 0, 0, 0},
+	};
+
+	while ((c = getopt_long(argc, argv, "c:p:o:s:e:dh", prog_opts, NULL))!= -1) {
+		switch (c) {
+			case 'c':
+				shim.container_id = strdup(optarg);
+				break;
+			case 'p':
+				shim.proxy_sock_fd = (int)parse_numeric_option(optarg);
+				if (shim.proxy_sock_fd < 0) {
+					err_exit("Invalid value for proxy socket fd\n");
+				}
+				break;
+			case 'o':
+				shim.proxy_io_fd = (int)parse_numeric_option(optarg);
+				if (shim.proxy_io_fd < 0) {
+					err_exit("Invalid value for proxy IO fd\n");
+				}
+				break;
+			case 's':
+				shim.io_seq_no = (uint64_t)parse_numeric_option(optarg);
+				if (shim.io_seq_no < 0) {
+					err_exit("Invalid value for I/O seqence no\n");
+				}
+				break;
+			case 'e':
+				shim.err_seq_no = (uint64_t)parse_numeric_option(optarg);
+				if (shim.err_seq_no < 0) {
+					err_exit("Invalid value for error sequence no\n");
+				}
+				break;
+			case 'd':
+				debug = true;
+				break;
+			case 'h':
+				print_usage();
+				exit(EXIT_SUCCESS);
+			default:
+				print_usage();
+				exit(EXIT_FAILURE);
+		}
+	}
+
+	if ( !shim.container_id) {
+		err_exit("Missing container id\n");
+	}
+
+	if ( shim.proxy_sock_fd == -1) {
+		err_exit("Missing proxy socket file descriptor\n");
+	}
+
+	if ( shim.proxy_io_fd == -1) {
+		err_exit("Missing proxy I/O file descriptor\n");
+	}
+
+	if (shim.io_seq_no == 0) {
```

Comment:
Probably worth explaining why `-e` is optional otherwise someone might see that as a bug and try to fix it ;)